src/zope/interface/interface.py

```
@@ -421,16 +540,124 @@ def get(self, name, default=None):
+    # It can't be a straight up ``@property`` descriptor, though, because accessing
+    # it on the class returns a ``property`` object, not the desired string.
+    #
+    # (3) Implement a data descriptor (``__get__`` and ``__set__``) that
+    # is both a string, and also does the redirect of ``__module__`` to ``__ibmodule__``
+    # and does the correct thing with the ``instance`` argument to ``__get__`` is None
+    # (returns the class's value.)
+    #
+    # This works, preserves the ability to read and write
+    # ``__module__``, and eliminates any penalty accessing other
+    # attributes. But it slows down accessing ``__module__`` of instances by 200%
+    # (40ns to 124ns).
+    #
+    # (4) As in the last step, but make it a non-data descriptor (no ``__set__``).
+    #
+    # If you then *also* store a copy of ``__ibmodule__`` in
+    # ``__module__`` in the instances dict, reading works for both
+    # class and instance and is full speed for instances. But the cost
+    # is storage space, and you can't write to it anymore, not without
+    # things getting out of sync.
+    #
+    # (Actually, ``__module__`` was never meant to be writable. Doing
+    # so would break BTrees and normal dictionaries, as well as the
+    # repr, maybe more.)
+    #
+    # That leaves us with a metaclass. Here we can have our cake and
+    # eat it too: no extra storage, and C-speed access to the
+    # underlying storage. The only cost is that metaclasses tend to
+    # make people's heads hurt. (But still less than the descriptor-is-string, I think.)
+
+    def __new__(cls, name, bases, attrs):
+        try:
+            # Figure out what module defined the interface.
+            # This is how cPython figures out the module of
+            # a class, but of course it does it in C. :-/
+            __module__ = sys._getframe(1).f_globals['__name__']
+        except (AttributeError, KeyError): # pragma: no cover
+            pass
```

So if this exception is caught and silently ignored...